Bluetooth: Host: Conn: Iterate over connections in TX list to avoid starvation of other connections

aahmed-dewinelabs opened the pull request:
Currently, starvation can happen when sending data through the `bt_conn_tx_processor` as connections who are currently not able to send are chosen again in the next iteration. 

By iterating over all the connections in the TX list, we can find another connection which is currently able to send data, instead of returning NULL when the first connection can't send data. This avoids starvation of other connections which are be able to send data (another connection could have LL buffers or TX context).

Discord discussion: https://discord.com/channels/720317445772017664/1326845512829370420

aahmed-dewinelabs committed:
bluetooth: host: iterate over connections in TX list

Iterate over the list of connections to find a connection
that is able to send data, instead of returning NULL if
the first connection can't send data.

This solves the problem with starvation of other connections.

Signed-off-by: Ahmed Ahmed <ahmed.ahmed@dewinelabs.com>

jhedberg commented:
@aahmed-dewinelabs thanks for the pull request! The change itself looks reasonable, but you need to fix your commit message to align with the contribution requirements: https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-message-guidelines

Hint: just look at the commit history (`git log <file>`) for the file you're modifying, and try to follow similar style for your commit message. The Compliance check has failed now and blocks this from being merged.

alwa-nordic commented on subsys/bluetooth/host/conn.c:
```diff
@@ -898,6 +898,22 @@ __maybe_unused static bool dont_have_methods(struct bt_conn *conn)
+	sys_snode_t *node = sys_slist_get(&bt_dev.le.conn_ready);
+	struct bt_conn *conn = CONTAINER_OF(node, struct bt_conn, _conn_ready);
+	(void)atomic_set(&conn->_conn_ready_lock, 0);
+	bt_conn_data_ready(conn);
```
I think this will result in a busy-loop when `cannot_send_to_controller` returns true for all ready connections. If so, can we avoid that?

aahmed-dewinelabs replied:
Would be just adding it back to the end of the list with append instead of `bt_conn_data_ready` be an option? This way we don't trigger `bt_tx_irq_raise`. 
I mean it would probably be better to overhaul the `get_conn_ready` function to try to always return a ready connection and only NULL if there are none instead of moving around in the list. I might try implementing something like this in the future.

alwa-nordic replied:
I think just moving it to the back of the list is not a sufficient fix for your initial problem. The initial problem is that there is some ISO data ready to send somewhere in the list and we want to dig it up. But if we just move the head to the back and `get_conn_ready` returns without calling `bt_tx_irq_raise`, the scheduler will sleep until something else calls `bt_tx_irq_raise`, and the ISO data stays delayed.

alwa-nordic replied:
A possible solution is to loop over the list (e.g. using `SYS_SLIST_FOR_EACH_CONTAINER_SAFE`) until something that can be sent is found, and `sys_slist_remove` that node.

aahmed-dewinelabs replied:
I like the idea, I will look into it. But why would you remove it then?

alwa-nordic replied:
I just meant `sys_slist_remove` instead of `sys_slist_get` because it's not necessarily at the head of the list. Nothing special. `get_conn_ready` will still re-add it to the back of the list if `should_stop_tx` returns `false`.

alwa-nordic replied:
Another solution is to have separate lists for ACL and ISO connections.

aahmed-dewinelabs replied:
Or when inserting into the list with `bt_conn_data_ready` we could make sure that ACL connections always come after ISO connections.

Thalley commented on subsys/bluetooth/host/conn.c:
```diff
@@ -898,6 +898,22 @@ __maybe_unused static bool dont_have_methods(struct bt_conn *conn)
+	/* Move reference to the end to allow for another connection to be
+	 * scheduled the next time the TX processor is called.
+	 */
+	if (sys_slist_len(&bt_dev.le.conn_ready) == 1) {
```
```suggestion
	if (sys_slist_peek_head(&bt_dev.le.conn_ready) == sys_slist_peek_tail(&bt_dev.le.conn_ready)) {
```

This matches the comment a bit better, and also only requires comparing two values (O(1)) instead of looping over the list (O(n))

Thalley commented on subsys/bluetooth/host/conn.c:
```diff
@@ -898,6 +898,22 @@ __maybe_unused static bool dont_have_methods(struct bt_conn *conn)
 		(conn->has_data == NULL);
 }
 
+static void move_head_to_tail(void)
```
Should calls to this function, or at least the slist access, but guarded by k_sched_lock? 

(reference to https://github.com/zephyrproject-rtos/zephyr/pull/83700)

cvinayak commented on subsys/bluetooth/host/conn.c:
```diff
@@ -956,8 +977,6 @@ struct bt_conn *get_conn_ready(void)
 			LOG_DBG("appending %p to back of TX queue", conn);
 			bt_conn_data_ready(conn);
 		}
-
-		return conn;
```
by removing this your are incrementing reference count for `conn` that is already referenced (present) in the `slist` .

aahmed-dewinelabs replied:
I currently only see that it is referenced inside `bt_conn_data_ready` when `conn->has_data` returns true. Where is it referenced when this is not the case?

aahmed-dewinelabs replied:
I think now I understand, we are removing it from the list but not decrementing the reference counter, therefore we continue to use that reference. In my opinion this is confusing behavior  - removing the reference and increasing  it when returning it is better I think.

jhedberg replied:
The reference count is supposed to reflect the number of pointer variables pointing at the same object at any given moment. When you call `sys_slist_get()` the pointer gets removed from the list, but since you don't need it for anything (since you already have another reference locally) the right thing to do would indeed be to drop this reference. The clean way to do that however would be with two separate pointer variables, even though they're pointing at the same object. In an ideal world I'd like our `unref()` APIs to take a pointer to a pointer and in addition to decrementing the reference count they'd also set the pointer NULL so that it can't be used anymore, however sthis would break a well established APIs so we can't really do that.

cvinayak reviewed (dismissed):
Will there be any effort to investigate the regression in multiple connections bsim CI tests? This is possible due to incorrect reference counting.

Thalley commented:
@aahmed-dewinelabs In general, please do not resolve reviewers comments yourself: https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#workflow-suggestions-that-help-reviewers

Thalley commented on subsys/bluetooth/host/conn.c:
```diff
@@ -900,67 +900,67 @@ __maybe_unused static bool dont_have_methods(struct bt_conn *conn)
+	sys_snode_t *prev = NULL;
 
 	if (dont_have_viewbufs()) {
-		/* We will get scheduled again when the (view) buffers are freed. If you
-		 * hit this a lot, try increasing `CONFIG_BT_CONN_FRAG_COUNT`
```
Minor: Why was the `you` moved?

aahmed-dewinelabs replied:
Good question, somehow the format has moved it.

Thalley commented on subsys/bluetooth/host/conn.c:
```diff
@@ -900,67 +900,67 @@ __maybe_unused static bool dont_have_methods(struct bt_conn *conn)
-		(void)atomic_set(&conn->_conn_ready_lock, 0);
+		if (should_stop_tx(conn)) {
+			/* Move reference off the list */
+			sys_slist_remove(&bt_dev.le.conn_ready, prev, &conn->_conn_ready);
```
Should we add some sanity checks as asserts?

e.g. `prev != conn` and possible also `tmp != conn` to ensure that we do not remove `prev` and that we don't get stuck in infinite loops. 

If the same conn somehow gets added twice to `bt_dev.le.conn_ready` then it the list item will point to itself, which will cause an infinite loop

aahmed-dewinelabs replied:
Sure. We could add `__ASSERT_NO_MSG(prev != &conn->_conn_ready && tmp != conn);` before `sys_slist_remove`?

Thalley replied:
> Sure. We could add `__ASSERT_NO_MSG(prev != &conn->_conn_ready && tmp != conn);` before `sys_slist_remove`?

I think `__ASSERT_NO_MSG(prev != conn)` before the remove, and `__ASSERT_NO_MSG(tmp != conn)` somewhere at the beginning of the foreach, as those two values should never be the same (as that would indicate that a conn was added twice to this list)

jhedberg commented on subsys/bluetooth/host/conn.c:
```diff
@@ -900,67 +900,67 @@ __maybe_unused static bool dont_have_methods(struct bt_conn *conn)
+			/* Move reference off the list */
+			sys_slist_remove(&bt_dev.le.conn_ready, prev, &conn->_conn_ready);
+			(void)atomic_set(&conn->_conn_ready_lock, 0);
+			bt_conn_unref(conn);
```
Sorry, I realize this is as it is now because of my misleading guidance. I didn't realize that the `conn` that you have in the loop is just from iterating a list, i.e. until you do `sys_slist_remove()` *the list* owns the reference and you don't. Once you remove from the list you own that reference, which means the when you do this `bt_conn_unref(conn)` you are no longer allowed to do anything with the `conn` pointer. So, the right thing to do would be to just keep the reference as is, and then you just `return conn;` without any additional `ref()` calls.

aahmed-dewinelabs replied:
Thanks for your feedback. Keeping track of the conn references can be a bit confusing.
I removed the `bt_conn_unref` and now return this reference in this if branch.

aahmed-dewinelabs committed:
bluetooth: host: add asserts to foreach in get_conn_ready

Assert that the previous connection in the list and the tmp element
in the loop are different to the current connection.
Otherwise, if the same connection was somehow added twice it
could result in an infinite loop.

Signed-off-by: Ahmed Ahmed <ahmed.ahmed@dewinelabs.com>

aahmed-dewinelabs committed:
bluetooth: host: fix connection reference before returning

As the connection was removed from the list, we now own its reference.
Keep this reference and return it to the TX processor.

Signed-off-by: Ahmed Ahmed <ahmed.ahmed@dewinelabs.com>
